### PR TITLE
autoconnect nbft systemd service

### PR DESCRIPTION
**Pull request by igaw**
Add support to auto connect via NBFT.

@mwilck during updating our downstream packages I noticed these patches never made it upstream. I think they should also be upstream. Any objections?

### CODE REVIEW

**Comment by mwilck** [comment]
No objections.

File: nvmf-autoconnect/systemd/nvmf-connect-nbft.service.in
Comment (martin-belanger) [inline nvmf-autoconnect/systemd/nvmf-connect-nbft.service.in:10]: Instead of using `ExecStartPre` to load the kernel module, I think that the preferred way is to do as follows under the `[Unit]` section:
```
[Unit]

Wants=modprobe@nvme_fabrics.service
After=modprobe@nvme_fabrics.service
```

Comment (igaw) [reply nvmf-autoconnect/systemd/nvmf-connect-nbft.service.in:10]: Sounds reasonable. Good catch! Thanks

**Review by hreinecke — COMMENTED** [review]
Argl. I _so_ hate the 'nbft' (and it's previous incarnation 'ibft') network device naming.
You always have the trouble to keep the udev rules in sync between initrd and the running system, not to mention the confusion arising when for some reason not booting from nbft
or nbft parsing fails after configuration.

File: nvmf-autoconnect/systemd/nvmf-connect-nbft.service.in
Comment (hreinecke) [reply nvmf-autoconnect/systemd/nvmf-connect-nbft.service.in:10]: Call me stupid, but where are the actual 'connect' calls?
Are they done by the 'nvmf-autoconnect.service'?
If so, shouldn't we add an explicit order like 'Before' here?
If not, how does autoconnect happen?

Comment (hreinecke) [reply nvmf-autoconnect/systemd/nvmf-connect-nbft.service.in:10]: And, incidentally, we're using 'ExecStartPre' in 'nvmf-autoconnect.in', too. Maybe we should be changing that, too.

**Comment by igaw** [comment]
> Argl. I _so_ hate the 'nbft' (and it's previous incarnation 'ibft') network device naming. You always have the trouble to keep the udev rules in sync between initrd and the running system, not to mention the confusion arising when for some reason not booting from nbft or nbft parsing fails after configuration.

Any ideas how to improve this then? I know with newer kernels a device can have alternative names. If we could make one of these 'stable' we don't have to do this udev rename dance.

**Comment by mwilck** [comment]
@hreinecke, we have been discussing this up and down in various places. My conclusion is that using a separate name space like `ibft$X` or `nbft$X` is the best thing we can currently do. Dracut's code for network interface naming would need to be completely rewritten if we were to use `eth$X` or system-style network interace names. It's actually a benefit that the names of the NBFT interfaces don't depend on the persistent network device naming rules in the root FS: We *don't need* to make sure that these rules are always in sync betwen root FS and initrd, *because* we use a separate name space.

See https://bugzilla.suse.com/show_bug.cgi?id=1194910#c45 and follow-up comments. 
If you read through that bug, you can see that the initial intention was to have dracut generate `eth$X` names that are consistent with the present udev rules, that we seriously tried to achieve this, and that we eventually abandoned the idea because we couldn't figure out how to make it work reliably.

Your point about "not booting from the NBFT" is valid. But TBH, if a system that's normally booted via NBFT now is booted differently, it's a different system. A common problem is the difference between installation time (possibly without NBFT) and runtime (with NBFT), but our NVMe-oF/TCP testing with the `nbft$X` approach has not revealed any issues, neither on VMs nor on actual hardware. The point is that wicked sees the interface configuration through its plugin, and that the udev rule in this PR makes sure no renaming happens.

Note also that the udev rule in this PR is much simpler than the rule-generator approach that we currently have for iSCSI. Changing iSCSI to use the same approach is on my todo list.

The current status quo, like it or not, is that dracut generates the `nbft$X` interface name, which is consistent with it's behavior for iSCSI, and has not been controversial in the Timberland group or upstream dracut PRs. If `nbft*` names exist, they shouldn't be changed. That's all what the udev rule ensures. If you or anyone else can come up with initrd code that takes the to-be-booted system's persistent interface naming rules into account in a reliable way, fine with me; in that case the rule from this PR will be a no-op because no `nbft$X` interfaces will exist any more.

**Comment by mwilck** [comment]
Obviously, upstream **nvme-cli** should not enforce any interface naming policy. But the udev rule in this PR doesn't. It just ensures that names matching a certain pattern, which is distinct from any other interface name patterns, shouldn't be changed. It's debatable whether such a rule should be part of upstream **nvme-cli**, but don't see a big problem with that.

File: nvmf-autoconnect/systemd/nvmf-connect-nbft.service.in
Comment (johnmeneghini) [inline nvmf-autoconnect/systemd/nvmf-connect-nbft.service.in:14]: Has any one tested this with nvme-fc boot?  We also support nvme-fc boot too.

**Review by johnmeneghini — COMMENTED** [review]
This looks like it should work.  NVMe/FC boot does rely on nvme-cli, but since the nvmfc-boot-connections.service runs first, and nvme connect-all doesn't currently default --nbft, this should work.

File: nvmf-autoconnect/systemd/nvmf-connect-nbft.service.in
Comment (johnmeneghini) [reply nvmf-autoconnect/systemd/nvmf-connect-nbft.service.in]: > Call me stupid, but where are the actual 'connect' calls? Are they done by the 'nvmf-autoconnect.service'? If so, shouldn't we add an explicit order like 'Before' here? If not, how does autoconnect happen?

The nvm connect-all --nbft call is currently done by dracut. 

https://github.com/dracutdevs/dracut/blob/6acfecae572fb457115b276b5b64d9424ad5187b/modules.d/95nvmf/nvmf-autoconnect.sh#L35

So maybe it's not a good idea to add it here.

File: nvmf-autoconnect/systemd/nvmf-connect-nbft.service.in
Comment (mwilck) [reply nvmf-autoconnect/systemd/nvmf-connect-nbft.service.in:14]: This command can be issued multiple times with no risk. The kernel notices if some of the requested connections are already established.

The point here is that during initrd processing, not all connections may have been brought up successfully, because not all network interface have been up. The typical usage scenario is multipath, where we don't want to get stuck in the initrd because one path is not available (yet). Thus  dracut will proceed mounting and switching root as soon as a single path is available for accessing the root FS. This, in turn, may cause dracut to switch "prematurely" if there are additional healthy paths, i.e. before it has brought up paths that can be brought up. For that reason, we need to make a 2nd bring-up attempt after switching root.

This way, we avoid being stuck in the initrd if paths are missing, while still bringing up all paths if all is good.

Comment (mwilck) [reply nvmf-autoconnect/systemd/nvmf-connect-nbft.service.in:14]: @johnmeneghini's comment makes me realize that this is the legacy syntax. It must be `nvme connect-all --nbft` for upstream.

**Review by mwilck — CHANGES_REQUESTED** [review]

File: nvmf-autoconnect/systemd/nvmf-connect-nbft.service.in
Comment (igaw) [reply nvmf-autoconnect/systemd/nvmf-connect-nbft.service.in]: Updated the nvme invocation. Anything else to do?

Comment (johnmeneghini) [reply nvmf-autoconnect/systemd/nvmf-connect-nbft.service.in]: I still have some concerns about how this will work together with dracut and network manager during  nbft boot. I've asked  @tbzatek to review this.  Can we add Tomas as a reviewer?

https://github.com/tbzatek

I'd also like to test these changes in my nvmft boot testbed before we merge. I'll do that next week.

**Review by johnmeneghini — COMMENTED** [review]
@pvalena and @tbzatek can you please review these changes?

**Comment by pvalena** [comment]
> @pvalena and @tbzatek can you please review these changes?

Looks good. I have no insight into nvme-cli, but FWIW I can see no conflict with what dracut does.

**Review by tbzatek — APPROVED** [review]
From a brief look it looks fine, I'm not a systemd expert though. I'm going to test it on a real environment but I don't see anything that should block this pull request.

**Comment by mwilck** [comment]
@johnmeneghini, wrt your concerns about NetworkManager: The autoconnect scripts are only concerned with the NVMe part of the connection establishment process.

Making another attempt to connect after switching root makes sense if dracut didn't bring up all connections during initrd processing, which can happen 

1. if some network connections failed / were unavailable during boot,
2. in multipath scenarios, when the rootfs was mounted before all connections were established.
 
In case 1., making a second autoconnect attempt after switching root will only have an effect if the network management tool (NetworkManager) had succeeded in bringing up network routes that dracut had failed to bring up. Whether that has any chance to succeed depends on the configuration. In case 2., autoconnect may succeed even with no post-switch-root changes in the network configuration.

Either way, this is orthogonal to any attempts to set up NBFT-defined network connections. The service starts after "network-online.target", and that's where NM comes into play. While working towards "network-online.target", NM might read the NBFT, see if all NBFT-defined network interfaces have been configured, and try to bring up any missing ones. I think we all understand that "network-online.target" is a rather clumsy all-or-nothing target which is a good fit for laptops that need a working internet connection, but less so for a server booting from NVMeoF in a complex networking environment.

**Comment by tbzatek** [comment]
Okay, this appears to be working fine on my simple test setup. Uptime 19 hours without any glitch.

**Comment by igaw** [comment]
Thanks!

**Pull request merged**
